From ffffd3a659c6c528daa312686c7532597db9d38d Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Cl=C3=A9ment=20Le=20Goffic?= <clement.legoffic@foss.st.com>
Date: Tue, 13 May 2025 15:41:17 +0200
Subject: [PATCH 835/841] fix(st-uart): aarch32: wait for UART ISR register TC
 bit for console flush
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Loop over the UART_ISR_TC bit to make sure the FIFO is emptied and all
the data is out of the shift register.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/arm-trusted-firmware.git 63af05e633ea872990f738a50718ce10192dfd9c fix(st-uart): aarch32: wait for UART ISR register TC bit for console flush]
Change-Id: Ie71df304225bf8b5f7d89eaa28b918d02c91d535
Signed-off-by: Clément Le Goffic <clement.legoffic@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/tf-a/+/457871
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Domain-Review: Amelie DELAUNAY <amelie.delaunay@foss.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: Maxime MERE <maxime.mere@foss.st.com>
---
 drivers/st/uart/aarch32/stm32_console.S | 11 ++++-------
 1 file changed, 4 insertions(+), 7 deletions(-)

diff --git a/drivers/st/uart/aarch32/stm32_console.S b/drivers/st/uart/aarch32/stm32_console.S
index 005dee9ac..1942c6bfb 100644
--- a/drivers/st/uart/aarch32/stm32_console.S
+++ b/drivers/st/uart/aarch32/stm32_console.S
@@ -219,7 +219,7 @@ endfunc console_stm32_core_getc
 	 *
 	 * In : r0 - console base address
 	 * Out : void.
-	 * Clobber list : r0, r1, r2
+	 * Clobber list : r0, r1
 	 * ---------------------------------------------------------------
 	 */
 func console_stm32_core_flush
@@ -231,14 +231,11 @@ func console_stm32_core_flush
 	ldr	r1, [r0, #USART_CR1]
 	tst	r1, #USART_CR1_UE
 	beq	1f
-	/* Check Transmit Data Register Empty */
-	mov	r2, #USART_TIMEOUT
-txe_loop_3:
-	subs	r2, r2, #1
-	beq	1f
+	/* Check Transmission Complete flag */
+tc_loop:
 	ldr	r1, [r0, #USART_ISR]
 	tst	r1, #USART_ISR_TXE
-	beq	txe_loop_3
+	beq	tc_loop
 1:
 	bx	lr
 endfunc console_stm32_core_flush
-- 
2.47.2

